app/scripts/controllers/app-state.test.js

```diff
@@ -29,5 +44,259 @@ describe('AppStateController', () => {
+
+      updateStateSpy.mockRestore();
+    });
+  });
+
+  describe('getUnlockPromise', () => {
+    it('waits for unlock if the extension is locked', async () => {
+      appStateController = createAppStateController();
+      const isUnlockedMock = jest
+        .spyOn(appStateController, 'isUnlocked')
+        .mockReturnValue(false);
+      const waitForUnlockSpy = jest.spyOn(appStateController, 'waitForUnlock');
+
+      appStateController.getUnlockPromise(true);
+      expect(isUnlockedMock).toHaveBeenCalled();
+      expect(waitForUnlockSpy).toHaveBeenCalledWith(expect.any(Function), true);
+    });
+
+    it('resolves immediately if the extension is already unlocked', async () => {
+      appStateController = createAppStateController();
+      const isUnlockedMock = jest
+        .spyOn(appStateController, 'isUnlocked')
+        .mockReturnValue(true);
+
+      await expect(
+        appStateController.getUnlockPromise(false),
+      ).resolves.toBeUndefined();
+
+      expect(isUnlockedMock).toHaveBeenCalled();
+    });
+  });
+
+  describe('waitForUnlock', () => {
+    it('resolves immediately if already unlocked', async () => {
+      const emitSpy = jest.spyOn(appStateController, 'emit');
+      const resolveFn = jest.fn();
+      appStateController.waitForUnlock(resolveFn, false);
+      expect(emitSpy).toHaveBeenCalledWith('updateBadge');
+      expect(appStateController._messenger.call).toHaveBeenCalledTimes(0);
+    });
+
+    it('creates approval request when waitForUnlock is called with shouldShowUnlockRequest as true', async () => {
+      jest.spyOn(appStateController, 'isUnlocked').mockReturnValue(false);
+
+      const resolveFn = jest.fn();
+      appStateController.waitForUnlock(resolveFn, true);
+
+      expect(appStateController._messenger.call).toHaveBeenCalledTimes(1);
+      expect(appStateController._messenger.call).toHaveBeenCalledWith(
+        'ApprovalController:addRequest',
+        expect.objectContaining({
+          id: expect.any(String),
+          origin: 'AppStateController',
+          type: 'unlock',
+        }),
+        true,
+      );
+    });
+  });
+
+  describe('handleUnlock', () => {
+    beforeEach(() => {
+      jest.spyOn(appStateController, 'isUnlocked').mockReturnValue(false);
+    });
+    it('accepts approval request revolving all the related promises', async () => {
+      const emitSpy = jest.spyOn(appStateController, 'emit');
+      const resolveFn = jest.fn();
+      appStateController.waitForUnlock(resolveFn, true);
+
+      appStateController.handleUnlock();
+
+      expect(emitSpy).toHaveBeenCalled();
+      expect(emitSpy).toHaveBeenCalledWith('updateBadge');
+      expect(appStateController._messenger.call).toHaveBeenCalled();
+      expect(appStateController._messenger.call).toHaveBeenCalledWith(
+        'ApprovalController:acceptRequest',
+        expect.any(String),
+      );
+    });
+    it('returns without call messenger if no approval request in pending', async () => {
```

Comment:
If we add a `try catch` to the accept, should we also have a unit test to verify if the `call` throws, the controller doesn't.

Reply:
yeap, once I've applied your suggestions above I also included a test case to log if rejecting an approval request throws.